ops/main.py

```diff
@@ -169,6 +169,22 @@ def main(charm_class):
     # TODO: For Windows, when symlinks are used, this is not a valid
     #       method of getting an event name (see LP: #1854505).
     juju_exec_path = Path(sys.argv[0])
+    has_dispatch = juju_exec_path.name == 'dispatch'
+    if has_dispatch:
+        # The executable was 'dispatch', which means the actual hook we want to
+        # run needs to be looked up in the JUJU_DISPATCH_PATH env var, where it
+        # should be a path relative to the charm directory (the directory that
+        # holds `dispatch`). If that path actually exists, we need to exec to
+        # it instead of continuing.
+        #
+        # We don't look at the existence of JUJU_DISPATCH_PATH because we don't
+        # unset it on exec'ing into the hook, and it could be a symlink back to
+        # ourselves.
+        dispatch_path = Path(os.environ['JUJU_DISPATCH_PATH'])
+        juju_exec_path = juju_exec_path.parent / dispatch_path
+        if juju_exec_path.exists():
+            sys.argv[0] = str(juju_exec_path)
+            os.execv(sys.argv[0], sys.argv)
```

Comment:
I don't think we want execv. We want subprocess.call.

eg, if 'hooks/config-changed' exists, the framework will call it, but we *also* trigger `Charm.on.config_changed`.

Reply:
I don't understand why we need to execute anything external at all.

For example: If JUJU_DISPATCH_PATH is set to `hooks/install`, for example, why we don't just call the method registered to `install`?

Wouldn't that be all?

Reply:
`hooks/install` could be a anything, e.g. a shell script